src/middlewared/middlewared/plugins/pool.py

```diff
@@ -2177,6 +2185,157 @@ def import_disk_msdosfs_locales(self):
 
         await self.middleware.call_hook('pool.post_export', pool=pool, options=options)
 
+    @item_method
+    @accepts(
+        Int("id"),
+        Dict(
+            "options",
+            Dict(
+                "geli",
+                Str("passphrase", private=True, default=""),
+            ),
+        )
+    )
+    @job(lock="pool_expand")
+    async def expand(self, job, id, options):
+        """
+        Expand pool to fit all available disk space.
+        """
+
+        pool = await self._get_instance(id)
+
+        if pool["encrypt"]:
+            if not pool["is_decrypted"]:
+                raise CallError("You can only expand decrypted pool")
+
+            for error in (await self.__pool_lock_pre_check(pool, options["geli"]["passphrase"])).errors:
+                raise CallError(error.errmsg)
+
+        try:
+            sysctl.filter("kern.geom.debugflags")[0].value = 16
+
+            geli_resize = []
+            try:
+                for vdev in sum(pool["topology"].values(), []):
+                    if vdev["type"] != "DISK":
+                        logger.debug("Not expanding vdev of type %r", vdev["type"])
+                        continue
+
+                    if vdev["status"] != "ONLINE":
+                        logger.debug("Not expanding vdev that is %r", vdev["status"])
+                        continue
+
+                    partition_number = RE_DISKPART.match(vdev["device"])
+                    if partition_number is None:
+                        logger.debug("Could not parse partition number from %r", vdev["device"])
+                        continue
+
+                    assert partition_number.group(1) == vdev["disk"]
+                    partition_number = int(partition_number.group(2)[1:])
+
+                    mediasize = bsd.geom.geom_by_name("LABEL", vdev["device"]).provider.mediasize
+
+                    await run("camcontrol", "reprobe", vdev["disk"])
+                    await run("gpart", "recover", vdev["disk"])
+                    await run("gpart", "resize", "-i", str(partition_number), vdev["disk"])
+
+                    if pool["encrypt"]:
+                        geli_resize_cmd = (
+                            "geli", "resize", "-s", str(mediasize), vdev["device"]
+                        )
+                        rollback_cmd = (
+                            "gpart", "resize", "-i", str(partition_number), "-s", str(mediasize), vdev["disk"]
+                        )
+
+                        logger.warning("It will be obligatory to notify GELI that the provider has been resized: %r",
+                                       join_commandline(geli_resize_cmd))
+                        logger.warning("Or to resize provider back: %r",
+                                       join_commandline(rollback_cmd))
```

Comment:
Whats the intent of logging about this?

Reply:
We need to run `geli resize` with old size of partition or the data is lost. That's why it's logged ASAP (in very rare case when middleware crashes).